modules/nextflow/src/main/groovy/nextflow/scm/AssetManager.groovy

```diff
@@ -94,39 +104,52 @@ class AssetManager {
     /**
      * Create a new asset manager with the specified pipeline name
      *
-     * @param pipeline The pipeline to be managed by this manager e.g. {@code nextflow-io/hello}
+     * @param pipelineName The pipeline to be managed by this manager e.g. {@code nextflow-io/hello}
+     * @param revision Revision ID for the selected pipeline (git branch, tag or commit SHA number)
      */
-    AssetManager( String pipelineName, HubOptions cliOpts = null) {
+    AssetManager( String pipelineName, String revision = null, HubOptions cliOpts = null) {
         assert pipelineName
         // read the default config file (if available)
         def config = ProviderConfig.getDefault()
         // build the object
-        build(pipelineName, config, cliOpts)
+        build(pipelineName, revision, config, cliOpts)
     }
 
-    AssetManager( String pipelineName, Map config ) {
+    AssetManager( String pipelineName, String revision, Map config) {
         assert pipelineName
         // build the object
-        build(pipelineName, config)
+        build(pipelineName, revision, config)
     }
 
     /**
      * Build the asset manager internal data structure
      *
      * @param pipelineName A project name or a project repository Git URL
+     * @param revision Revision ID for the selected pipeline (git branch, tag or commit SHA number)
      * @param config A {@link Map} holding the configuration properties defined in the {@link ProviderConfig#DEFAULT_SCM_FILE} file
      * @param cliOpts User credentials provided on the command line. See {@link HubOptions} trait
      * @return The {@link AssetManager} object itself
      */
     @PackageScope
-    AssetManager build( String pipelineName, Map config = null, HubOptions cliOpts = null ) {
+    AssetManager build( String pipelineName, String revision = null, Map config = null, HubOptions cliOpts = null ) {
+
+        // if requested revision corresponds to the default branch, then unset it
+        // this avoids duplication of the default branch
+        if ( revision ) {
+            def referenceManager = new AssetManager(pipelineName, null, cliOpts)
+            if ( revision == referenceManager.getDefaultBranch() ) {
+                revision = null
+            }
```

Comment:
A nested `AssetManager` creation smell very bad. Why not calling directly `this.getDefaultBranch()` ?

Reply:
To use `this.getDefaultBranch` we still need to mock a number of things, so that we can then get the manifest from the provider (or at least, this is the best solution I could find):
```
        this.providerConfigs = ProviderConfig.createFromMap(config)

        // these four are temporary, just to get the default branch
        this.project = resolveName(pipelineName, null)
        this.localPath = checkProjectDir(this.project, null)
        this.hub = checkHubProvider(cliOpts)
        this.provider = createHubProvider(hub)

        // if requested revision corresponds to the default branch, then unset it
        // this avoids duplication of the default branch
        if ( revision && revision == this.getDefaultBranch() ) {
            revision = null
        }

        // now the proper initialisation
        this.revision = revision
        this.project = resolveName(pipelineName, this.revision)
        this.localPath = checkProjectDir(this.project, this.revision)
        this.hub = checkHubProvider(cliOpts)
        this.provider = createHubProvider(hub)
        this.provider.setRevision(this.revision)

        [...]
```

I went for the nested creation in the PR for conciseness basically. Happy to move to the alternate solution above. What are your thoughts @pditommaso ?